File: rust-resampler/src/smote_knn.rs
use kdtree::KdTree;
use kdtree::distance::squared_euclidean;
use ndarray::{Array1, Array2, Axis};
use ndarray_rand::RandomExt;
use ndarray_rand::rand_distr::Uniform;
use rand::{Rng, thread_rng};
use rayon::prelude::*;
use std::error::Error;

#[derive(Debug, Clone)]
enum Metric {
    Euclidean,
    Manhattan,
    // 추가 metric 지원 가능
}

impl Metric {
    fn distance(&self, a: &Array1<f64>, b: &Array1<f64>) -> f64 {
        match self {
            Metric::Euclidean => (a - b).mapv(|x| x.powi(2)).sum().sqrt(),
            Metric::Manhattan => (a - b).mapv(|x| x.abs()).sum(),
        }
    }
}

fn compute_knn(
    x: &Array2<f64>,
    k: usize,
    metric: Metric,
) -> Result<(Array2<f64>, Array2<usize>), Box<dyn Error>> {
    if x.shape()[0] <= k {
        return Err(format!("샘플 수({})가 k({})보다 작거나 같습니다!", x.shape()[0], k).into());
    }

    let dimensions = x.shape()[1];
    let mut kdtree = KdTree::new(dimensions);
    for (i, row) in x.outer_iter().enumerate() {
        kdtree.add(row.to_vec(), i)?;
    }

    let mut distances = Array2::zeros((x.shape()[0], k));
    let mut indices = Array2::zeros((x.shape()[0], k));

    x.outer_iter()
        .into_par_iter()
        .enumerate()
        .for_each(|(i, row)| {
            let query = row.to_vec();
            let knn = kdtree.nearest(&query, k + 1, &squared_euclidean).unwrap(); // +1 for self
            for (j, (dist, &idx)) in knn.iter().skip(1).enumerate() {
                // skip self
                distances[[i, j]] = dist.sqrt();
                indices[[i, j]] = idx;
            }
        });

    Ok((distances, indices))
}

fn find_knn_for_sample(
    x: &Array2<f64>,
    sample_idx: usize,
    k: usize,
    metric: Metric,
) -> Result<(Array1<f64>, Array1<usize>), Box<dyn Error>> {
    if sample_idx >= x.shape()[0] {
        return Err(format!(
            "샘플 인덱스({})가 데이터 크기({})를 초과합니다!",
            sample_idx,
            x.shape()[0]
        )
        .into());
    }
    if x.shape()[0] <= k {
        return Err(format!("샘플 수({})가 k({})보다 작거나 같습니다!", x.shape()[0], k).into());
    }

    let row = x.row(sample_idx).to_owned();
    let mut distances = Array1::zeros(k);
    let mut indices = Array1::zeros(k);

    // 전체 pairwise 계산 (작은 데이터셋 가정)
    let dists: Vec<(f64, usize)> = x
        .outer_iter()
        .enumerate()
        .map(|(i, r)| (metric.distance(&row, &r.to_owned()), i))
        .filter(|&(_, i)| i != sample_idx)
        .collect();

    let mut sorted_dists = dists;
    sorted_dists.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap());

    for (j, (d, idx)) in sorted_dists.iter().take(k).enumerate() {
        distances[j] = *d;
        indices[j] = *idx;
    }

    Ok((distances, indices))
}

fn compute_pairwise_distances(x: &Array2<f64>, metric: Metric) -> Array2<f64> {
    let n = x.shape()[0];
    let mut dists = Array2::zeros((n, n));
    x.outer_iter()
        .into_par_iter()
        .enumerate()
        .for_each(|(i, row_i)| {
            for (j, row_j) in x.outer_iter().enumerate() {
                if i != j {
                    dists[[i, j]] = metric.distance(&row_i, &row_j);
                }
            }
        });
    dists
}

fn smote(
    x: &Array2<f64>,
    y: &Array1<i64>,
    k: usize,
    sampling_ratio: f64,
    random_state: Option<u64>,
    metric: Metric,
) -> Result<(Array2<f64>, Array1<i64>), Box<dyn Error>> {
    let mut rng = thread_rng();
    if let Some(seed) = random_state {
        // rand crate에서 seeded RNG 지원 (추가 구현 필요)
        // 여기서는 간단히 무시하고 thread_rng 사용
    }

    // 클래스 카운트
    let mut class_counts: HashMap<i64, usize> = HashMap::new();
    for &label in y.iter() {
        *class_counts.entry(label).or_insert(0) += 1;
    }

    if class_counts.len() != 2 {
        return Err("SMOTE는 이진 분류 문제에만 적용 가능합니다!".into());
    }

    let minority_class = class_counts
        .iter()
        .min_by_key(|&(_, count)| count)
        .unwrap()
        .0
        .clone();
    let minority_indices: Vec<usize> = y
        .iter()
        .enumerate()
        .filter(|&(_, &l)| l == minority_class)
        .map(|(i, _)| i)
        .collect();
    let minority_x = x.select(Axis(0), &minority_indices);

    let n_minority = minority_indices.len();
    let n_majority = x.shape()[0] - n_minority;
    let n_synthetic = ((n_majority - n_minority) as f64 * sampling_ratio) as usize;

    if n_synthetic == 0 {
        return Ok((x.clone(), y.clone()));
    }

    let (_, indices) = compute_knn(&minority_x, k, metric.clone())?;

    let mut synthetic_samples: Vec<Array1<f64>> = Vec::with_capacity(n_synthetic);
    let mut synthetic_labels: Vec<i64> = Vec::with_capacity(n_synthetic);

    for _ in 0..n_synthetic {
        let sample_idx = rng.gen_range(0..n_minority);
        let neighbor_idx = rng.gen_range(0..k);
        let neighbor_sample_idx = indices[[sample_idx, neighbor_idx]];

        let alpha = rng.gen_range(0.0..1.0);
        let synth = &minority_x.row(sample_idx)
            + alpha * (&minority_x.row(neighbor_sample_idx) - &minority_x.row(sample_idx));

        synthetic_samples.push(synth.to_owned());
        synthetic_labels.push(minority_class);
    }

    let mut x_resampled = x.clone();
    for sample in synthetic_samples {
        x_resampled.push_row(sample.view())?;
    }
    let mut y_resampled = y.clone();
    y_resampled.append(&mut Array1::from_vec(synthetic_labels));

    Ok((x_resampled, y_resampled))
}

// smote_with_custom_knn은 smote와 유사하므로, metric을 smote에 통합하여 중복 구현 생략 (필요 시 확장)

fn main() -> Result<(), Box<dyn Error>> {
    // 예시 사용
    let x = Array2::random((100, 10), Uniform::new(0., 10.));
    let mut y = Array1::zeros(100);
    y.slice_mut(s![50..]).fill(1); // 불균형 클래스

    let (x_res, y_res) = smote(&x, &y, 5, 1.0, None, Metric::Euclidean)?;
    println!("Resampled shape: {:?}, {:?}", x_res.shape(), y_res.shape());

    Ok(())
}

File: poseidon/data/smote_knn.py
"""
SMOTE (Synthetic Minority Over-sampling Technique) 및 KNN (K-Nearest Neighbors) 구현 모듈

이 모듈은 불균형 데이터셋에서 소수 클래스의 합성 샘플을 생성하기 위한 SMOTE 알고리즘과
그에 필요한 KNN 계산 함수들을 제공합니다.

주요 기능:
- KNN 계산: compute_knn, find_knn_for_sample, compute_pairwise_distances
- SMOTE 구현: smote, smote_with_custom_knn
- 이진 분류 문제에서 클래스 불균형 해결
- 다양한 거리 측정 방법 지원 (유클리드, 맨하탄 등)
"""

import numpy as np
from sklearn.neighbors import NearestNeighbors
from sklearn.metrics.pairwise import pairwise_distances
from tqdm import tqdm


def compute_knn(X, k=5, metric="euclidean"):
    """
    K-최근접 이웃(K-Nearest Neighbors)을 계산하는 함수

    Parameters:
    -----------
    X : array-like, shape (n_samples, n_features)
        입력 데이터
    k : int, default=5
        고려할 최근접 이웃의 수
    metric : str, default='euclidean'
        거리 측정 방법 ('euclidean', 'manhattan', 'minkowski' 등)

    Returns:
    --------
    distances : array, shape (n_samples, k)
        각 샘플에서 k개의 최근접 이웃까지의 거리
    indices : array, shape (n_samples, k)
        각 샘플의 k개의 최근접 이웃 인덱스
    """
    if X.shape[0] <= k:
        raise ValueError(f"샘플 수({X.shape[0]})가 k({k})보다 작거나 같습니다!")

    # NearestNeighbors 모델 생성 및 학습
    nn = NearestNeighbors(n_neighbors=k + 1, metric=metric)  # +1은 자기 자신 포함
    nn.fit(X)

    # 각 점에서 k+1개의 최근접 이웃 찾기 (자기 자신 포함)
    distances, indices = nn.kneighbors(X)

    # 자기 자신 제거 (첫 번째 열이 자기 자신)
    distances = distances[:, 1:]  # 거리에서 자기 자신 제거
    indices = indices[:, 1:]  # 인덱스에서 자기 자신 제거

    return distances, indices


def find_knn_for_sample(X, sample_idx, k=5, metric="euclidean"):
    """
    특정 샘플에 대한 K-최근접 이웃을 찾는 함수

    Parameters:
    -----------
    X : array-like, shape (n_samples, n_features)
        입력 데이터
    sample_idx : int
        KNN을 찾을 샘플의 인덱스
    k : int, default=5
        고려할 최근접 이웃의 수
    metric : str, default='euclidean'
        거리 측정 방법

    Returns:
    --------
    distances : array, shape (k,)
        해당 샘플에서 k개의 최근접 이웃까지의 거리
    indices : array, shape (k,)
        해당 샘플의 k개의 최근접 이웃 인덱스
    """
    if sample_idx >= X.shape[0]:
        raise ValueError(
            f"샘플 인덱스({sample_idx})가 데이터 크기({X.shape[0]})를 초과합니다!"
        )

    if X.shape[0] <= k:
        raise ValueError(f"샘플 수({X.shape[0]})가 k({k})보다 작거나 같습니다!")

    # NearestNeighbors 모델 생성 및 학습
    nn = NearestNeighbors(n_neighbors=k + 1, metric=metric)
    nn.fit(X)

    # 특정 샘플에 대한 최근접 이웃 찾기
    distances, indices = nn.kneighbors([X[sample_idx]])

    # 자기 자신 제거
    distances = distances[0, 1:]  # 첫 번째 행, 자기 자신 제외
    indices = indices[0, 1:]  # 첫 번째 행, 자기 자신 제외

    return distances, indices


def compute_pairwise_distances(X, metric="euclidean"):
    """
    모든 샘플 쌍 간의 거리를 계산하는 함수

    Parameters:
    -----------
    X : array-like, shape (n_samples, n_features)
        입력 데이터
    metric : str, default='euclidean'
        거리 측정 방법

    Returns:
    --------
    distances : array, shape (n_samples, n_samples)
        모든 샘플 쌍 간의 거리 행렬
    """

    return pairwise_distances(X, metric=metric)


def smote(X, y, k=5, sampling_ratio=1.0, random_state=None):
    """
    SMOTE (Synthetic Minority Over-sampling Technique) 알고리즘을 구현하는 함수

    Parameters:
    -----------
    X : array-like, shape (n_samples, n_features)
        입력 특성 데이터
    y : array-like, shape (n_samples,)
        타겟 레이블
    k : int, default=5
        SMOTE에서 사용할 최근접 이웃의 수
    sampling_ratio : float, default=1.0
        소수 클래스에 생성할 합성 샘플의 비율 (1.0 = 100%)
    random_state : int, optional
        재현 가능한 결과를 위한 랜덤 시드

    Returns:
    --------
    X_resampled : array, shape (n_samples + n_synthetic, n_features)
        오버샘플링된 특성 데이터
    y_resampled : array, shape (n_samples + n_synthetic,)
        오버샘플링된 타겟 레이블
    """
    if random_state is not None:
        np.random.seed(random_state)

    X = np.array(X)
    y = np.array(y)

    # 클래스별 샘플 수 계산
    unique_classes, class_counts = np.unique(y, return_counts=True)

    if len(unique_classes) != 2:
        raise ValueError("SMOTE는 이진 분류 문제에만 적용 가능합니다!")

    # 소수 클래스 식별
    minority_class = unique_classes[np.argmin(class_counts)]

    # 소수 클래스 샘플 추출
    minority_indices = np.where(y == minority_class)[0]
    minority_X = X[minority_indices]

    # 생성할 합성 샘플 수 계산
    n_minority = len(minority_indices)
    n_majority = len(X) - n_minority
    n_synthetic = int((n_majority - n_minority) * sampling_ratio)

    if n_synthetic <= 0:
        return X, y

    # 소수 클래스에 대한 KNN 계산
    _, indices = compute_knn(minority_X, k=k)

    # 합성 샘플 생성
    synthetic_samples = []
    synthetic_labels = []

    for _ in tqdm(range(n_synthetic), desc="SMOTE 합성 샘플 생성", unit="샘플"):
        # 랜덤하게 소수 클래스 샘플 선택
        sample_idx = np.random.randint(0, n_minority)

        # 해당 샘플의 k개 최근접 이웃 중 하나를 랜덤 선택
        neighbor_idx = np.random.randint(0, k)
        neighbor_sample_idx = indices[sample_idx, neighbor_idx]

        # 합성 샘플 생성 (선형 보간)
        alpha = np.random.random()  # 0과 1 사이의 랜덤 값
        synthetic_sample = minority_X[sample_idx] + alpha * (
            minority_X[neighbor_sample_idx] - minority_X[sample_idx]
        )

        synthetic_samples.append(synthetic_sample)
        synthetic_labels.append(minority_class)

    # 원본 데이터와 합성 샘플 결합
    X_resampled = np.vstack([X, np.array(synthetic_samples)])
    y_resampled = np.hstack([y, np.array(synthetic_labels)])

    return X_resampled, y_resampled


def smote_with_custom_knn(
    X, y, k=5, sampling_ratio=1.0, random_state=None, metric="euclidean"
):
    """
    사용자 정의 KNN 함수를 사용하는 SMOTE 구현

    Parameters:
    -----------
    X : array-like, shape (n_samples, n_features)
        입력 특성 데이터
    y : array-like, shape (n_samples,)
        타겟 레이블
    k : int, default=5
        SMOTE에서 사용할 최근접 이웃의 수
    sampling_ratio : float, default=1.0
        소수 클래스에 생성할 합성 샘플의 비율
    random_state : int, optional
        재현 가능한 결과를 위한 랜덤 시드
    metric : str, default="euclidean"
        거리 측정 방법

    Returns:
    --------
    X_resampled : array, shape (n_samples + n_synthetic, n_features)
        오버샘플링된 특성 데이터
    y_resampled : array, shape (n_samples + n_synthetic,)
        오버샘플링된 타겟 레이블
    """
    if random_state is not None:
        np.random.seed(random_state)

    X = np.array(X)
    y = np.array(y)

    # 클래스별 샘플 수 계산
    unique_classes, class_counts = np.unique(y, return_counts=True)

    if len(unique_classes) != 2:
        raise ValueError("SMOTE는 이진 분류 문제에만 적용 가능합니다!")

    # 소수 클래스 식별
    minority_class = unique_classes[np.argmin(class_counts)]

    # 소수 클래스 샘플 추출
    minority_indices = np.where(y == minority_class)[0]
    minority_X = X[minority_indices]

    # 생성할 합성 샘플 수 계산
    n_minority = len(minority_indices)
    n_majority = len(X) - n_minority
    n_synthetic = int((n_majority - n_minority) * sampling_ratio)

    if n_synthetic <= 0:
        return X, y

    # 소수 클래스에 대한 KNN 계산 (사용자 정의 함수 사용)
    _, indices = compute_knn(minority_X, k=k, metric=metric)

    # 합성 샘플 생성
    synthetic_samples = []
    synthetic_labels = []

    for _ in tqdm(range(n_synthetic), desc="SMOTE 합성 샘플 생성", unit="샘플"):
        # 랜덤하게 소수 클래스 샘플 선택
        sample_idx = np.random.randint(0, n_minority)

        # 해당 샘플의 k개 최근접 이웃 중 하나를 랜덤 선택
        neighbor_idx = np.random.randint(0, k)
        neighbor_sample_idx = indices[sample_idx, neighbor_idx]

        # 합성 샘플 생성 (선형 보간)
        alpha = np.random.random()  # 0과 1 사이의 랜덤 값
        synthetic_sample = minority_X[sample_idx] + alpha * (
            minority_X[neighbor_sample_idx] - minority_X[sample_idx]
        )

        synthetic_samples.append(synthetic_sample)
        synthetic_labels.append(minority_class)

    # 원본 데이터와 합성 샘플 결합
    X_resampled = np.vstack([X, np.array(synthetic_samples)])
    y_resampled = np.hstack([y, np.array(synthetic_labels)])

    return X_resampled, y_resampled

__all__ = [
    'compute_knn',
    'smote',
]
